### PR TITLE
Fix error code received from vttablets on context expiration

### PR DESCRIPTION
**Pull request by GuptaManan100**
## Description

This PR fixes the error codes returned when context expires on vttablet even before the execution of the query starts. Previously we would return the error by calling `fmt.Errorf`, which on being translated to grpc error codes, would result in `Code_UNKNOWN`. Instead if we use `vterrors.Errorf` and use `vtrpcpc.Code_CANCELLED`, we can get a better error code.

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vttablet/tabletserver/connpool/dbconn_test.go [hunk 1]
```diff
@@ -479,18 +448,14 @@ func TestDBNoPoolConnKill(t *testing.T) {
...
 	db.AddQuery(query, &sqltypes.Result{})
 	// Kill failed because we are not able to connect to the database
 	db.EnableConnFail()
 	err = dbConn.Kill("test kill", 0)
-	want := "errno 2013"
-	if err == nil || !strings.Contains(err.Error(), want) {
-		t.Errorf("Exec: %v, want %s", err, want)
-	}
+	require.Error(t, err)
+	require.ErrorContains(t, err, "errno 2013")
```
Comment (mattlord) [inline go/vt/vttablet/tabletserver/connpool/dbconn_test.go:458]: Things like this are probably good places to test our MySQL error handling as this should be `sqlerror.CRServerLost`, no?

**Review by mattlord — APPROVED** [review]

File: go/vt/vttablet/tabletserver/connpool/dbconn_test.go [hunk 2]
```diff
@@ -212,15 +195,11 @@ func TestDBConnDeadline(t *testing.T) {
...
-	}
+	require.NoError(t, err)
 
 	_, err = dbConn.Exec(ctx, sql, 1, false)
-	want := "context deadline exceeded before execution started"
-	if err == nil || !strings.Contains(err.Error(), want) {
-		t.Errorf("Exec: %v, want %s", err, want)
-	}
+	require.Error(t, err)
+	require.ErrorContains(t, err, "(errno 3024) (sqlstate HY000): Query execution was interrupted, maximum statement execution time exceeded, before execution started")
```
Comment (harshit-gangal) [inline go/vt/vttablet/tabletserver/connpool/dbconn_test.go:203]: Let us change the last line to `before executing on underlying database`.

Comment (deepthi) [reply go/vt/vttablet/tabletserver/connpool/dbconn_test.go:203]: I think the current message is fine. Why "underlying database"?? I'd just drop the comma between exceeded and before.

File: go/vt/vttablet/tabletserver/connpool/dbconn.go [hunk 3]
```diff
@@ -166,7 +166,7 @@ func (dbc *Conn) execOnce(ctx context.Context, query string, maxrows int, wantfi
 	// Check if the context is already past its deadline before
 	// trying to execute the query.
 	if err := ctx.Err(); err != nil {
-		return nil, fmt.Errorf("%v before execution started", err)
+		return nil, vterrors.Errorf(vtrpcpb.Code_CANCELED, "%s, before execution started", dbc.getErrorMessageFromContextError(ctx))
```
Comment (deepthi) [inline go/vt/vttablet/tabletserver/connpool/dbconn.go:169]:
```suggestion
		return nil, vterrors.Errorf(vtrpcpb.Code_CANCELED, "%s before execution started", dbc.getErrorMessageFromContextError(ctx))
```

File: go/vt/vttablet/tabletserver/connpool/dbconn.go [hunk 4]
```diff
@@ -229,7 +235,7 @@ func (dbc *Conn) FetchNext(ctx context.Context, maxrows int, wantfields bool) (*
 	// Check if the context is already past its deadline before
 	// trying to fetch the next result.
 	if err := ctx.Err(); err != nil {
-		return nil, fmt.Errorf("%v before reading next result set", err)
+		return nil, vterrors.Errorf(vtrpcpb.Code_CANCELED, "%s, before reading next result set", dbc.getErrorMessageFromContextError(ctx))
```
Comment (deepthi) [inline go/vt/vttablet/tabletserver/connpool/dbconn.go:238]:
```suggestion
		return nil, vterrors.Errorf(vtrpcpb.Code_CANCELED, "%s before reading next result set", dbc.getErrorMessageFromContextError(ctx))
```

**Review by deepthi — COMMENTED** [review]
LGTM, but other reviews should be addressed.

File: go/vt/vttablet/tabletserver/connpool/dbconn_test.go (hunk 2 above)
Comment (deepthi) [reply go/vt/vttablet/tabletserver/connpool/dbconn_test.go:203]: Left this as suggestions on the appropriate LOC.

File: go/vt/vttablet/tabletserver/connpool/dbconn_test.go (hunk 1 above)
Comment (GuptaManan100) [reply go/vt/vttablet/tabletserver/connpool/dbconn_test.go]: Fixed

File: go/vt/vttablet/tabletserver/connpool/dbconn.go (hunk 3 above)
Comment (GuptaManan100) [reply go/vt/vttablet/tabletserver/connpool/dbconn.go]: Fixed

File: go/vt/vttablet/tabletserver/connpool/dbconn.go (hunk 4 above)
Comment (GuptaManan100) [reply go/vt/vttablet/tabletserver/connpool/dbconn.go]: Fixed

File: go/vt/vttablet/tabletserver/connpool/dbconn_test.go (hunk 2 above)
Comment (harshit-gangal) [reply go/vt/vttablet/tabletserver/connpool/dbconn_test.go]: I see that execution started at the VTGate and VTTablet, but not on MySQL, so adding that message would make it clear.

**Review by harshit-gangal — APPROVED** [review]

**Pull request merged**